modules/stochastic_tools/src/reporters/AdaptiveMonteCarloDecision.C

```diff
@@ -0,0 +1,106 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdaptiveMonteCarloDecision.h"
+#include "Sampler.h"
+
+registerMooseObject("StochasticToolsApp", AdaptiveMonteCarloDecision);
+
+InputParameters
+AdaptiveMonteCarloDecision::validParams()
+{
+  InputParameters params = GeneralReporter::validParams();
+  params.addClassDescription("Generic reporter which decides whether or not to accept a proposed "
+                             "sample in Adaptive Monte Carlo type of algorithms.");
+  params.addRequiredParam<ReporterName>("output_value",
+                                        "Value of the model output from the SubApp.");
+  params.addParam<ReporterValueName>(
+      "output_required",
+      "output_required",
+      "Modified value of the model output from this reporter class.");
+  params.addParam<ReporterValueName>("inputs", "inputs", "Uncertain inputs to the model.");
+  params.addRequiredParam<SamplerName>("sampler", "The sampler object.");
+  return params;
+}
+
+AdaptiveMonteCarloDecision::AdaptiveMonteCarloDecision(const InputParameters & parameters)
+  : GeneralReporter(parameters),
+    _output_value(&getReporterValue<std::vector<Real>>("output_value")),
+    _output_required(declareValueByName<Real>("output_required")),
+    _inputs(declareValueByName<std::vector<Real>>("inputs")),
+    _step(getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")->timeStep()),
+    _sampler(getSampler("sampler"))
+{
+
+  _ais = dynamic_cast<const AdaptiveImportanceSampler *>(&_sampler);
+
+  // Initialize the required variables depending upon the type of adaptive Monte Carlo algorithm
+  if (_ais)
+  {
+    _inputs.resize(_sampler.getNumberOfCols());
+    _prev_val.resize(_sampler.getNumberOfCols());
+    for (unsigned int i = 0; i < _sampler.getNumberOfCols(); ++i)
+      _prev_val[i] = _ais->getInitialValues()[i];
+    _prev_val_out = 1.0;
+  }
+}
```

Comment:
It seems like there should be an error if we aren't using an `AdaptiveImportanceSampler` as such, here is a suggestion:

```suggestion
AdaptiveMonteCarloDecision::AdaptiveMonteCarloDecision(const InputParameters & parameters)
  : GeneralReporter(parameters),
    _output_value(getReporterValue<std::vector<Real>>("output_value")),
    _output_required(declareValue<Real>("output_required")),
    _inputs(declareValue<std::vector<Real>>("inputs")),
    _step(_fe_problem.timeStep()),
    _ais(getSampler<AdaptiveImportanceSampler>("sampler")),
{
  _inputs.resize(_ais.getNumberOfCols());
  _prev_val = _ais.getInitialValues();
  _prev_val_out = 1.0;
}
```

`getSampler<AdaptiveImportanceSampler>` will error if the inputted sampler is not the right type.

Reply:
Also, what is `_output_required` being used for?